# Data Cleaning-DSLR

The reviews of different Laptops, Smart Phones, Headphones, Smart Watches, DSLR (Professional Cameras), Printers, Monitors, Home Theaters, Routers from Flipkart.
The data preprocessing will be done on each of the dataset seaprately. First we will filter out the data to get equal number of reviews for each rating. The heading of the review is also extracted as the heading also can help in determining the rating.

In [29]:
#import the dataset
import pandas as pd
import numpy as np
pd.set_option('Display.max_columns',None)
pd.set_option('Display.max_rows',None)

df=pd.read_csv("DSLR Rating.csv")
df.head()

Unnamed: 0  Rating                    Heading  \
0           0     5.0                  Brilliant   
1           1     4.0  Good choice for the money   
2           2     5.0                  Wonderful   
3           3     NaN                        NaN   
4           4     NaN                        NaN   

                                              Review Product  
0  It's a nice budget entry level mirrorless came...    DSLR  
1  For a beginner, this Camera seems to be the be...    DSLR  
2  it's a awesome camera loved it.\nreally underr...    DSLR  
3                                                NaN    DSLR  
4                                                NaN    DSLR

#### Observations:
* The feature unnamed is index. Hence we can drop this feature.

In [30]:
df.drop(['Unnamed: 0'], axis=1, inplace=True)
df.head()

Rating                    Heading  \
0     5.0                  Brilliant   
1     4.0  Good choice for the money   
2     5.0                  Wonderful   
3     NaN                        NaN   
4     NaN                        NaN   

                                              Review Product  
0  It's a nice budget entry level mirrorless came...    DSLR  
1  For a beginner, this Camera seems to be the be...    DSLR  
2  it's a awesome camera loved it.\nreally underr...    DSLR  
3                                                NaN    DSLR  
4                                                NaN    DSLR

### Exploratory Data Analysis

In [31]:
#check the dimensions of the data (Headphone)
df.shape

(5729, 4)

* The dataset has 5729 rows and 4 columns
* The dataset has 1 label - 'Rating' and 3 features

In [32]:
#check the names of columns in dataset
df.columns

Index(['Rating', 'Heading', 'Review', 'Product'], dtype='object')

In [33]:
#check the datatype of each feature
df.dtypes

Rating     float64
Heading     object
Review      object
Product     object
dtype: object

#### Observations:
   * All the feratures are of "object" data type.

In [34]:
#checking if there are any null values in the dataset
df.isna().sum()

Rating     118
Heading    118
Review     118
Product      0
dtype: int64

In [35]:
df[df.isna()]

Rating Heading Review Product
0        NaN     NaN    NaN     NaN
1        NaN     NaN    NaN     NaN
2        NaN     NaN    NaN     NaN
3        NaN     NaN    NaN     NaN
4        NaN     NaN    NaN     NaN
5        NaN     NaN    NaN     NaN
6        NaN     NaN    NaN     NaN
7        NaN     NaN    NaN     NaN
8        NaN     NaN    NaN     NaN
9        NaN     NaN    NaN     NaN
10       NaN     NaN    NaN     NaN
11       NaN     NaN    NaN     NaN
12       NaN     NaN    NaN     NaN
13       NaN     NaN    NaN     NaN
14       NaN     NaN    NaN     NaN
15       NaN     NaN    NaN     NaN
16       NaN     NaN    NaN     NaN
17       NaN     NaN    NaN     NaN
18       NaN     NaN    NaN     NaN
19       NaN     NaN    NaN     NaN
20       NaN     NaN    NaN     NaN
21       NaN     NaN    NaN     NaN
22       NaN     NaN    NaN     NaN
23       NaN     NaN    NaN     NaN
24       NaN     NaN    NaN     NaN
25       NaN     NaN    NaN     NaN
26       NaN     NaN    NaN     NaN
27       NaN     NaN    NaN     NaN
28       NaN     NaN    NaN     NaN
29       NaN     NaN    NaN     NaN
30       NaN     NaN    NaN     NaN
31       NaN     NaN    NaN     NaN
32       NaN     NaN    NaN     NaN
33       NaN     NaN    NaN     NaN
34       NaN     NaN    NaN     NaN
35       NaN     NaN    NaN     NaN
36       NaN     NaN    NaN     NaN
37       NaN     NaN    NaN     NaN
38       NaN     NaN    NaN     NaN
39       NaN     NaN    NaN     NaN
40       NaN     NaN    NaN     NaN
41       NaN     NaN    NaN     NaN
42       NaN     NaN    NaN     NaN
43       NaN     NaN    NaN     NaN
44       NaN     NaN    NaN     NaN
45       NaN     NaN    NaN     NaN
46       NaN     NaN    NaN     NaN
47       NaN     NaN    NaN     NaN
48       NaN     NaN    NaN     NaN
49       NaN     NaN    NaN     NaN
50       NaN     NaN    NaN     NaN
51       NaN     NaN    NaN     NaN
52       NaN     NaN    NaN     NaN
53       NaN     NaN    NaN     NaN
54       NaN     NaN    NaN     NaN
55       NaN     NaN    NaN     NaN
56       NaN     NaN    NaN     NaN
57       NaN     NaN    NaN     NaN
58       NaN     NaN    NaN     NaN
59       NaN     NaN    NaN     NaN
60       NaN     NaN    NaN     NaN
61       NaN     NaN    NaN     NaN
62       NaN     NaN    NaN     NaN
63       NaN     NaN    NaN     NaN
64       NaN     NaN    NaN     NaN
65       NaN     NaN    NaN     NaN
66       NaN     NaN    NaN     NaN
67       NaN     NaN    NaN     NaN
68       NaN     NaN    NaN     NaN
69       NaN     NaN    NaN     NaN
70       NaN     NaN    NaN     NaN
71       NaN     NaN    NaN     NaN
72       NaN     NaN    NaN     NaN
73       NaN     NaN    NaN     NaN
74       NaN     NaN    NaN     NaN
75       NaN     NaN    NaN     NaN
76       NaN     NaN    NaN     NaN
77       NaN     NaN    NaN     NaN
78       NaN     NaN    NaN     NaN
79       NaN     NaN    NaN     NaN
80       NaN     NaN    NaN     NaN
81       NaN     NaN    NaN     NaN
82       NaN     NaN    NaN     NaN
83       NaN     NaN    NaN     NaN
84       NaN     NaN    NaN     NaN
85       NaN     NaN    NaN     NaN
86       NaN     NaN    NaN     NaN
87       NaN     NaN    NaN     NaN
88       NaN     NaN    NaN     NaN
89       NaN     NaN    NaN     NaN
90       NaN     NaN    NaN     NaN
91       NaN     NaN    NaN     NaN
92       NaN     NaN    NaN     NaN
93       NaN     NaN    NaN     NaN
94       NaN     NaN    NaN     NaN
95       NaN     NaN    NaN     NaN
96       NaN     NaN    NaN     NaN
97       NaN     NaN    NaN     NaN
98       NaN     NaN    NaN     NaN
99       NaN     NaN    NaN     NaN
100      NaN     NaN    NaN     NaN
101      NaN     NaN    NaN     NaN
102      NaN     NaN    NaN     NaN
103      NaN     NaN    NaN     NaN
104      NaN     NaN    NaN     NaN
105      NaN     NaN    NaN     NaN
106      NaN     NaN    NaN     NaN
107      NaN     NaN    NaN     NaN
108      NaN     NaN    NaN     NaN
109      NaN     NaN    NaN     NaN
110      N

In [36]:
#dropping all numm values
df.dropna(inplace=True)

In [37]:
#cross checking null values
df.isna().sum()

Rating     0
Heading    0
Review     0
Product    0
dtype: int64

In [38]:
df.shape

(5611, 4)

#### Observations:
* There are 5611 rows in the dataset

In [39]:
df['Rating'].value_counts()

5.0    3934
4.0    1155
3.0     252
1.0     212
2.0      58
Name: Rating, dtype: int64

#### Observations:
* The labels are inbalanced. 
* The efficiency of review classifier will be better when we have equal or near equal number of reviews for each rating

#### Action:
* We will have as many number of reviews for each rating as there are in the 1 rating (least rating)
* The excess reviews dropped.
* Before dropping the excess reviews we will first check the length of reviews and drop long reviews as it will use up more space.

In [40]:
#check info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5611 entries, 0 to 5728
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Rating   5611 non-null   float64
 1   Heading  5611 non-null   object 
 2   Review   5611 non-null   object 
 3   Product  5611 non-null   object 
dtypes: float64(1), object(3)
memory usage: 219.2+ KB


#### Observations
   * The info() method thus returns the data type as well as the non-null values and memory usage.
   * Out of the total of 4 columns 3 columns are "object" type while rest of them are float datatype.

In [41]:
#check number of unique values in each class;
df.nunique()

Rating        5
Heading      95
Review     4097
Product       1
dtype: int64

#### Observations:
* The label rating has 5 unique values: 1, 2, 3, 4, 5
* The headings can be duplicate as it is kind of summary of the review.
* The reviews should be unique. Hence, we will drop the duplicate reviews to avoid over-fitting.

In [42]:
df.drop_duplicates(subset='Review', inplace=True)

In [43]:
#cross checking for diplicacy of reviews
print(df.shape)
print(df.nunique())

(4097, 4)
Rating        5
Heading      94
Review     4097
Product       1
dtype: int64


In [44]:
#checking the length of review
df['Review_word_counter']=df['Review'].str.strip().str.len()
df.head()

Rating                    Heading  \
0      5.0                  Brilliant   
1      4.0  Good choice for the money   
2      5.0                  Wonderful   
13     5.0        Best in the market!   
14     4.0                Good choice   

                                               Review Product  \
0   It's a nice budget entry level mirrorless came...    DSLR   
1   For a beginner, this Camera seems to be the be...    DSLR   
2   it's a awesome camera loved it.\nreally underr...    DSLR   
13  People who ever plan to progress from novice t...    DSLR   
14  Good Product. Great build. Decent battery life...    DSLR   

    Review_word_counter  
0                   433  
1                   508  
2                   346  
13                  183  
14                   82

The maximum number of reviews are for 5 star rating. We will sort the rating in ascending order and length of review. So, that we drop excess number of reviews for 5 star rating and get near balanced dataset.

In [45]:
sorted_df=df.sort_values(by=['Rating','Review_word_counter'])
sorted_df.head()

Rating                 Heading Review Product  Review_word_counter
1000     1.0        Terrible product    Bad    DSLR                    3
2360     1.0         Useless product    bad    DSLR                    3
3393     1.0  Don't waste your money   Miss    DSLR                    4
743      1.0         Useless product  Worst    DSLR                    5
2640     1.0               Very poor  Waste    DSLR                    5

In [46]:
sorted_df.head(100)

Rating                                          Heading  \
1000     1.0                                 Terrible product   
2360     1.0                                  Useless product   
3393     1.0                           Don't waste your money   
743      1.0                                  Useless product   
2640     1.0                                        Very poor   
612      1.0                           Not recommended at all   
3026     1.0                                         Horrible   
3433     1.0                           Worst experience ever!   
3804     1.0                                   Unsatisfactory   
3909     1.0                                   Unsatisfactory   
4955     1.0                        Did not meet expectations   
3846     1.0                                        Worthless   
4381     1.0                        Did not meet expectations   
1214     1.0                                        Worthless   
2450     1.0                                  Useless product   
3611     1.0                                         Horrible   
982      1.0                                        Worthless   
2354     1.0                        Did not meet expectations   
2413     1.0                                 Terrible product   
3680     1.0                           Don't waste your money   
3712     1.0                                  Useless product   
3912     1.0                                        Hated it!   
3913     1.0                        Did not meet expectations   
4903     1.0                                  Useless product   
2173     1.0                                        Hated it!   
2449     1.0                                        Worthless   
2736     1.0                                         Horrible   
2741     1.0                                   Unsatisfactory   
2976     1.0                           Not recommended at all   
3867     1.0                                Absolute rubbish!   
5000     1.0                                        Worthless   
561      1.0                                  Waste of money!   
1181     1.0                                Absolute rubbish!   
3582     1.0                                  Useless product   
5041     1.0                                  Waste of money!   
831      1.0                        Did not meet expectations   
2142     1.0                        Did not meet expectations   
2582     1.0                                   Unsatisfactory   
2733     1.0                                        Very poor   
3652     1.0                           Not recommended at all   
1343     1.0                           Worst experience ever!   
5267     1.0                                        Very poor   
2460     1.0                           Worst experience ever!   
3726     1.0                                  Waste of money!   
4448     1.0                                         Horrible   
2316     1.0                             Utterly Disappointed   
3555     1.0                             Utterly Disappointed   
2850     1.0                           Not recommended at all   
2907     1.0                           Worst experience ever!   
3464     1.0                             Utterly Disappointed   
4893     1.0                                         CHEATING   
2472     1.0                                  Waste of money!   
2357     1.0                           Worst experience ever!   
3908     1.0                           Worst experience ever!   
4548     1.0                                         Horrible   
686      1.0                           Don't waste your money   
2312     1.0                           Worst experience ever!   
1284     1.0                                         Horrible   
3667     1.0                                        Hated it!   
1970     1.0                                Absolute rubbish!   
4253     1.0                             

In [47]:
sorted_df['Rating'].value_counts()

5.0    2840
4.0     806
3.0     204
1.0     194
2.0      53
Name: Rating, dtype: int64

The least number of reviews are for rating '2'. We will make the number of reviews for rating '4' and '5' equal.

In [48]:
df_DSLR=pd.concat([df_1,df_2,df_3,df_4,df_5[:806]])
df_DSLR.head()

Rating                 Heading    Review Product  Review_word_counter
3393     1.0  Don't waste your money      Miss    DSLR                    4
743      1.0         Useless product     Worst    DSLR                    5
612      1.0  Not recommended at all  Very bad    DSLR                    8
3433     1.0  Worst experience ever!  Very bed    DSLR                    8
3909     1.0          Unsatisfactory  lite wet    DSLR                    8

In [49]:
df_DSLR.shape

(1977, 5)

In [50]:
#Reshuffling and reindexing the data
from sklearn.utils import shuffle
df_DSLR=shuffle(df_DSLR)
df_DSLR.reset_index(inplace=True,drop=True)
df_DSLR.head(100)

Rating                                            Heading  \
0      5.0                                          Must buy!   
1      4.0                                        Really Nice   
2      2.0                                           Moderate   
3      1.0                                          Worthless   
4      3.0                                          Just okay   
5      5.0                                  Terrific purchase   
6      1.0                             Worst experience ever!   
7      4.0                                    Value-for-money   
8      5.0                                  Worth every penny   
9      3.0                                       Does the job   
10     5.0                                          Brilliant   
11     4.0                                    Value-for-money   
12     3.0                                               Fair   
13     4.0                                          Very Good   
14     4.0                                          Very Good   
15     1.0                                          Worthless   
16     5.0                                          Fabulous!   
17     4.0                                    Value-for-money   
18     5.0                                Best in the market!   
19     5.0                                          Wonderful   
20     3.0                                          Just okay   
21     5.0                                 Highly recommended   
22     1.0                                          Hated it!   
23     1.0                             Worst experience ever!   
24     4.0                                          Very Good   
25     5.0                                          Brilliant   
26     5.0                                           Terrific   
27     4.0                                    Value-for-money   
28     4.0                                        Good choice   
29     4.0                                          Wonderful   
30     1.0                                    Useless product   
31     4.0                                         Delightful   
32     5.0                                 Highly recommended   
33     5.0                                           Terrific   
34     4.0                               Good quality product   
35     5.0                                  Terrific purchase   
36     1.0                             Don't waste your money   
37     5.0                                          Wonderful   
38     4.0                                        Really Nice   
39     4.0                                        Pretty good   
40     4.0                                         Delightful   
41     5.0                                  Terrific purchase   
42     5.0                                          Brilliant   
43     4.0                               Good quality product   
44     4.0                               Good quality product   
45     5.0                                     Classy product   
46     5.0                                Best in the market!   
47     1.0                               Utterly Disappointed   
48     4.0                                          Wonderful   
49     4.0                                          Very Good   
50     5.0                                 Highly recommended   
51     5.0                                 Highly recommended   
52     1.0                                    Useless product   
53     5.0                                          Excellent   
54     3.0                                     Decent product   
55     5.0                                     Classy product   
56     4.0  Great buy at this price range ..good for entry...   
57     5.0                              Mind-blowing purchase   
58     3.0                                               Fair   
59     1.0                                           Horrible   
60     4.0                             

In [51]:
df_DSLR.drop(['Review_word_counter'],axis=1,inplace=True)
df_DSLR.head()

Rating      Heading                                             Review  \
0     5.0    Must buy!                                awesome product....   
1     4.0  Really Nice  very good for this price range... beginners wi...   
2     2.0     Moderate                    not as expected picture quality   
3     1.0    Worthless  So satisfying with the product and delivery se...   
4     3.0    Just okay  The build quality is ok .....plastic but looks...   

  Product  
0    DSLR  
1    DSLR  
2    DSLR  
3    DSLR  
4    DSLR

In [52]:
df_DSLR.to_csv('DSLR Rating_cleaned.csv')

In [53]:
df_DSLR.shape

(1977, 4)